### ISSUE TITLE
Avoid raising S310 if user explicitly checks for URL scheme

### ISSUE THREAD

**Issue by 93578237** [astral-sh/ruff#7918]
* A minimal code snippet that reproduces the bug.

```python
import urllib.request


def foo(url: str) -> None:
    if not url.startswith(("https://", "http://")):
        raise ValueError
    req = urllib.request.Request(
        url,
        headers=headers,
        data=data,
        method=method,
    )
    resp = urllib.request.urlopen(req)
```
Also it would be great if ruff could recognize asserts (like `assert url.startswith(("https://", "http://"))`)
* The command you invoked (e.g., `ruff /path/to/file.py --fix`), ideally including the `--isolated` flag.

ruff --isolated --select S t.py
* The current Ruff settings (any relevant sections from your `pyproject.toml`).

No pyproject.toml
* The current Ruff version (`ruff --version`).

ruff 0.0.292

**Comment by qdegraaf** [comment]
Could you clarify why `urllib.request.Request` should be flagged? [Bandit's blacklist](https://bandit.readthedocs.io/en/latest/blacklists/blacklist_calls.html#b310-urllib-urlopen) does not include it as an issue.

**Comment by qdegraaf** [comment]
Ah I misunderstood. I thought from your description that the rule was not triggering and it should. But if I understand you correctly now, it is being flagged, and you want to change rule S310 to not flag if the URL passed to it has been checked to start with `https://` or `http://` and not say `file://`. This would not be specific to just `urllib.request.Request` I take it but all `urlopen` calls flagged by S310. 

@charliermarsh is that something that is desirable to implement or is this something that should be manually marked as a check skip by the dev in question. Docs just suggest:

```
/// To mitigate this risk, audit all uses of URL open functions and ensure that
/// only permitted schemes are used (e.g., allowing `http:` and `https:` and
/// disallowing `file:` and `ftp:`).
```

**Comment by charliermarsh** [comment]
I think this will be hard to get right with absolute certainty and so I'm somewhat hesitant to invest in it given that it's a security-related rule. If someone wants to try, though, I am happy to review it.

**Comment by charliermarsh** [comment]
We now no longer flag this if you use a string literal, which is at least an improvement.

**Comment by henryiii** [comment]
What about string literal but not in place? This is currently triggering the warning:

```python
url = "https://pypi.org/pypi?:action=list_classifiers"
context = ssl.create_default_context()
with urlopen(url, context=context) as response:
```

**Comment by Hawk777** [comment]
A string literal passed through `Request` is also flagged when it ideally shouldn’t be:
```python
import urllib

urllib.request.urlopen(urllib.request.Request("https://example.com/"))
```

(obviously in this case I could just pass the string literal directly; in reality, I’m constructing the `Request` because I have extra headers to add)

**Comment by charliermarsh** [comment]
I can fix a few of these.

**Comment by retu2libc** [comment]
It would be cool if stuff like
```Python
url = f"https://{self.target_url}{self.path}"
```
didn't trigger this.

**Comment by ViniGodoy** [comment]
I'll upvote this one. I also had to add the noqa, even implementing the same correction described in the [S310 "use instead" documentation](https://docs.astral.sh/ruff/rules/suspicious-url-open-usage/):

```python
def fetch(url: str, data=None, headers=None, method=None) -> HTTPResponse | HTTPError:
    headers = headers or {}
    if not url.startswith(("http:", "https:")):
        raise ValueError("URL must start with 'http:' or 'https:'")
    req = Request(url=url, data=data, headers=headers or {}, method=method)  # noqa: S310
    _log.info(f"--> FETCH {url}")
    try:
        return urllib.request.urlopen(req)  # noqa: S310
    except HTTPError as e:
        return e
```

I understand it could not understand some url concatenations, or if this check was done in a separate function. But at least he same use case described in the documentation should ideally be covered.

**Comment by dimaqq** [comment]
This one's a bit of a nit.

Folks who get the url from settings somewhere must slap `#noqa` on the call.

That's fine at that point, surely a reasonable developer also added some url schema validation.

But what happens in a year or two? Since `#noqa` is there and cannot be removed, any refactoring may break or what the validation and none would notice.

**Comment by MichaReiser** [comment]
Yeah, I think we could do better here. E.g. I find it surprising that the rule still raises an error even for the *Use instead* example in the documentation.

**Comment by hunterhogan** [comment]
So, here's the deal:

1. I don't have training or experience as a professional programmer.
2. I very much trust Ruff. I have been blown away by the quality of the linting and by the quality of explanatory documentation.

It naturally follows: I defer to Ruff's technical explanations.

[Bandit first flagged this](https://github.com/hunterhogan/mapFolding/issues/13) issue five months ago. I started using Ruff at least 10 weeks ago. Since I started using Ruff, I have read the [S310 documentation page](https://docs.astral.sh/ruff/rules/suspicious-url-open-usage/) at least 10 times. It is the only diagnostic error in that package that I can't fix. I've spent many hours trying to fix it. I literally copied and pasted the "Use instead" from the website and put it in my code. 

I kept looking for what I was doing wrong. Today, however, after a few hours and at least three copy-paste attempts, I decided to create an "Issue" and ask what I was doing wrong. I searched the open Issues and found this one.

Y'all don't need to improve the code. I like the linting rule, and I have not disabled it because I want to ensure I catch this issue. Please add to the documentation, however, that the diagnostic is a permanent warning. 

I greatly appreciate your hard work, and I do not think you need to put hard work into this issue.

**Comment by ethanstoner** [comment]
Measured this on ruff 0.16.5 to see what the rule actually keys on, because the thread has drifted between two different questions.

S310's only exemption is that the URL is a **string literal**. No scheme check of any shape affects it:

| snippet | S310 |
|---|---|
| literal URL, no guard at all | 0 |
| literal URL + `startswith` guard (the docs' *Use instead* example) | 0 |
| non-literal, no guard | 1 |
| non-literal + `if url.startswith("https://"):` | 1 |
| non-literal + `if not url.startswith(...): raise` | 1 |
| non-literal + `if not url.startswith(...): return` | 1 |
| non-literal + `assert url.startswith(...)` | 1 |

All with `ruff check --isolated --select S310`.

Two things follow.

**@MichaReiser's point from 2025-02-23 is already fixed** — the *Use instead* example from the S310 documentation passes cleanly now. But it passes because `url = "https://www.example.com/"` is a literal, not because the `startswith` check does anything. Deleting the `if` from that example leaves it equally clean.

That makes the documentation actively misleading: it presents a scheme check as the remedy for S310, and the scheme check is not what silences it. Anyone following the documented fix on a URL that came from a parameter or config — which is the only situation where the rule fires — gets the warning anyway. That is what @ViniGodoy and @dimaqq are describing above.

So there are two separable pieces of work here:

1. **Documentation.** Make the *Use instead* example show what actually silences the rule, or note that the guard does not. Small and low-risk, and it stops the loop of people reporting the documented fix not working.
2. **Recognising a guard.** The feature this issue asks for. @charliermarsh's 2023 hesitancy about doing this "with absolute certainty" on a security rule still applies, and a guard check would have to decide how far to follow the value.

I am happy to do (1) now, and (2) only if you want it and agree the scope first — a narrow version limited to a guard in the same function that dominates the call, bailing out on anything it cannot see, would be the version I would propose. Which of those, if either, would you take?
